### PR TITLE
Ensure compliant publisher API

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says. Connected to ros2/rmw#252.

### CODE REVIEW

File: rmw_connext_cpp/src/rmw_publisher.cpp [hunk 1]
```diff
@@ -441,80 +427,71 @@ rmw_return_loaned_message_from_publisher(
...
-    ConnextPublisherListener * pub_listener = publisher_info->listener_;
-    if (pub_listener) {
-      RMW_TRY_DESTRUCTOR(
-        pub_listener->~ConnextPublisherListener(),
-        ConnextPublisherListener, return RMW_RET_ERROR)
-      rmw_free(pub_listener);
-    }
+  ConnextPublisherListener * pub_listener = publisher_info->listener_;
 
+  if (!rmw_error_is_set()) {
```
Comment (hidmic) [inline rmw_connext_cpp/src/rmw_publisher.cpp:469]: Same as in other equivalent PRs, I don't like having to do this to ensure the first error that occurs is the one that pops up and that no subsequent error goes unnoticed.

**Comment by ivanpauno** [comment]
This PR has conflicts with https://github.com/ros2/rmw_connext/pull/444, the one that's later merged will need a rebase.

**Comment by hidmic** [comment]
@ros-pull-request-builder please retest this.

**Comment by hidmic** [comment]
CI up `test_rmw_implementation` against all RMW implementations:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11630/)

**Comment by hidmic** [comment]
> It would be good to double check locally with valgrind that this PR doesn't introduce any new leak.

Good point.

**Comment by hidmic** [comment]
I don't see any _new_ leaks, though `valgrind` isn't entirely happy about RTI Connext internals.

**Comment by jacobperron** [comment]
Looks like this PR introduced test failures in `rcl`: https://ci.ros2.org/view/nightly/job/nightly_linux_release/1623/testReport/

I've tried re-running tests with this PR reverted and it looks okay: https://ci.ros2.org/job/ci_linux/11705/

@hidmic Can you take a look at either fixing the tests or going forward with a revert in the meantime (https://github.com/ros2/rmw_connext/pull/448).

**Comment by hidmic** [comment]
Thanks for reporting @jacobperron. #449 should've taken care.

**Pull request merged**